landoapi/transplants.py

```diff
@@ -495,43 +606,330 @@ def check_landing_blockers(
+    assessment = TransplantAssessment()
+
+    # Run stack-level blocker checks.
+    for block in stack_blockers:
+        reason = block(transplant_state=transplant_state)
+        if reason is not None:
+            assessment.blockers.append(reason)
+
+    # Get the appropriate list of pairs to run checks against.
+    revision_check_pairs = transplant_state.revision_check_pairs()
+
+    # Run revision-level blockers checks.
+    for revision, diff in revision_check_pairs:
+        phid = revision["phid"]
+        for block in revision_blockers:
+            reason = block(
+                revision=revision,
+                diff=diff,
+                transplant_state=transplant_state,
+            )
+            if reason is not None:
+                assessment.blockers.append(reason)
+
+                if phid is not None and phid in transplant_state.landable_stack:
+                    transplant_state.landable_stack.remove_node(phid)
+                    transplant_state.stack.nodes[phid]["blocked"].append(reason)
+
+    # Run revision-level warning checks.
+    for revision, diff in revision_check_pairs:
+        for check in revision_warnings:
+            reason = check(
+                revision=revision, diff=diff, transplant_state=transplant_state
+            )
+
+            if reason is not None:
+                assessment.warnings.append(reason)
+
+    return assessment
+
+
+def assess_transplant_request(
+    phab: PhabricatorClient,
+    supported_repos: dict[str, Repo],
     stack_data: RevisionData,
+    stack: RevisionStack,
+    relman_group_phid: str,
     data_policy_review_phid: str,
-):
-    """Build all transplant blocker checks that need extra Phabricator data"""
-    assert all((isinstance(r, Repo) for r in repositories.values()))
-
-    return DEFAULT_OTHER_BLOCKER_CHECKS + [
-        # Configure uplift check with extra data.
-        check_uplift_approval(relman_group_phid, repositories, stack_data),
-        # Configure data policy check with the project PHID.
-        check_revision_data_classification(data_policy_review_phid),
-    ]
+    landing_assessment: Optional[LandingAssessmentState] = None,
+) -> tuple[TransplantAssessment, TransplantAssessmentState]:
+    """Assess the transplant request."""
+    # Get data from Lando.
+    landable_repos = get_landable_repos_for_revision_data(stack_data, supported_repos)
+
+    involved_phids = set()
+    reviewers = {}
+    for revision in stack_data.revisions.values():
+        involved_phids.update(gather_involved_phids(revision))
+        reviewers[revision["phid"]] = get_collated_reviewers(revision)
+
+    # Get more Phabricator data.
+    involved_phids = list(involved_phids)
+    users = user_search(phab, involved_phids)
+    projects = project_search(phab, involved_phids)
+
+    secure_project_phid = get_secure_project_phid(phab)
+    testing_tag_project_phids = get_testing_tag_project_phids(phab)
+    testing_policy_phid = get_testing_policy_phid(phab)
```

Comment:
I think this should also be done elsewhere, perhaps in the same place as relman_group_phid and data_policy_review_phid. Maybe they can all be put in a dataclass somewhere and fetched directly from there. Slightly out of scope, but at least we should not be introducing more of these calls in these methods.

Reply:
These calls were already a part of `assess_transplant_request`, but they were all happening in the `check_landing_warnings` function call. They've just been moved to an earlier location since we need them to build the `TransplantAssessmentState`. I'm going to leave them as-is for now, because what I would like to do is come up with a unified API for "retrieve and store all the required project PHIDs we need to inspect in Lando", and move them all into that API in a separate PR.

Reply:
Is there a bug tracking this work?

Reply:
No, it's just an idea at the moment. The main goal of the refactor was to have a more consistent interface for blockers/warnings than the scattered state we were in before. Cleaning up the scattered calls to Phabricator is something I want to do as well, but it's out of scope for this PR. I can file a bug for this if need be, it's just not a blocker for hg->git and can probably be done when we have some spare cycles.

Reply:
Filed [bug 1893498](https://bugzilla.mozilla.org/show_bug.cgi?id=1893498) for this.